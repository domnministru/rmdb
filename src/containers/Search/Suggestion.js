import React from "react";

import {DEFAULT_IMG_URL, SMALL_IMG} from "../../const";

import {Image} from "semantic-ui-react";
import noImage from "../../images/no-img.png"

const Suggestion = props => {
    const {title, release, path, type } = props;
    const imageLink = DEFAULT_IMG_URL + SMALL_IMG + path;

    return (
        <div className="suggestion-body">
            <Image className="suggestion-image"
                   src={
                       path ?
                           (imageLink) :
                           (noImage)
                   }>
            </Image>
            <div className="suggestion-info">
                <div className="suggestion-title">
                    <h2>{title}</h2>
                </div>
                <div className="suggestion-year">
                    <h4>{release}</h4>
                </div>
            </div>
            <div className="suggestion-type">
                {type}
            </div>
        </div>
    );
};

export default Suggestion;