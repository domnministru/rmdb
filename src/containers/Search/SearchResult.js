import React, {Component} from "react";
import {connect} from 'react-redux';
import OutsideClickHandler from 'react-outside-click-handler';

import {setSearchResultsVisibility} from './modules/searchAction';

import Suggestion from "./Suggestion";
import "../../styles/layout/_search.scss"

class SearchResult extends Component {
    render() {
        const {result, setSearchResultsVisibility} = this.props;

        return (
            <OutsideClickHandler onOutsideClick={() => setSearchResultsVisibility(false)}>
                <div className="results-area">
                    {
                        result.map(suggestion => {
                            return (
                                <Suggestion
                                    key={suggestion.id}
                                    title={suggestion.title || suggestion.name}
                                    release={suggestion.release_date || suggestion.first_air_date}
                                    type={suggestion.media_type}
                                    path={suggestion.poster_path || suggestion.profile_path}
                                />
                            )
                        })
                    }
                </div>
            </OutsideClickHandler>
        );
    }
}

const mapDispatchToProps = {
    setSearchResultsVisibility,
};

export default connect(null, mapDispatchToProps)(SearchResult);