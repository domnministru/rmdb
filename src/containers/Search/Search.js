import React, {Component} from 'react';
import {Container, Icon} from 'semantic-ui-react';
import {connect} from 'react-redux';

import {searchAll, setSearchResultsVisibility} from './modules/searchAction';

import SearchResult from './SearchResult';
import Spinner from "../../components/Spinner"
import "../../styles/layout/_search.scss"

export class Search extends Component {
    state = {query: null};

    handleInputChange = (e) => {
        this.setState({query: e.target.value})
    };

    handleSubmit = (e) => {
        e.preventDefault();
        if(this.state.query) {
            this.props.searchAll(this.state.query);
            this.props.setSearchResultsVisibility(true);
        }
    };

    render() {
        const {error, loading, result, isSearchResultsVisible} = this.props;
        const filteredResult = result.filter(item => (item.name || item.title));
        if (error) console.log(error);
        if (loading) return <Spinner/>;

        return (
            <div className="search-area">
                <Container className="primary-container">
                    <form className="search-form" onSubmit={this.handleSubmit}>
                        <Icon name="search"
                              size="large"
                              className="search-icon"
                        />
                        <input type="text"
                               className="search-input"
                               placeholder="Search for a Movie, Tv Show or Person"
                               onChange={this.handleInputChange}
                        />
                    </form>
                    {
                        isSearchResultsVisible &&
                        <SearchResult result={filteredResult} />
                    }
                </Container>
            </div>
        )
    };
}

const mapStateToProps = state => ({
    result: state.suggestions.suggestions,
    loading: state.suggestions.loading,
    error: state.suggestions.error,
    isSearchResultsVisible: state.suggestions.isSearchResultsVisible,
});

const mapDispatchToProps = {
    searchAll,
    setSearchResultsVisibility,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);