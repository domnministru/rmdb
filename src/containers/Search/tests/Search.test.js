import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Search } from '../Search';

it('Should be rendered correctly', () => {
    const wrapper = shallow(<Search result={[]} />);
    expect(wrapper.find('.search_area')).toHaveLength(1);
});

it('should match the snapshot', () => {
    const wrapper = shallow(<Search result={[]} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should match the snapshot with loading true', () => {
    const wrapper = shallow(<Search result={[]} loading={true} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should match the snapshot with error', () => {
    const wrapper = shallow(<Search result={[]} error={true} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should match the snapshot with isSearchResultsVisible', () => {
    const wrapper = shallow(<Search result={[]} isSearchResultsVisible={true} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should call a function on input change', () => {
    const wrapper = shallow(<Search result={[]} />);
    const mockCallback = jest.fn();
    wrapper.find('.search_form').simulate('submit');
    expect(mockCallback).toBeCalled();
});