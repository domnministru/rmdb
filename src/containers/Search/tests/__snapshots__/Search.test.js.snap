// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="search_area"
>
  <Container
    className="primary-container"
  >
    <form
      className="search_form"
      onSubmit={[Function]}
    >
      <Icon
        as="i"
        className="search_icon"
        name="search"
        size="large"
      />
      <input
        className="search_input"
        onChange={[Function]}
        placeholder="Search for a Movie, Tv Show or Person"
        type="text"
      />
    </form>
  </Container>
</div>
`;

exports[`should match the snapshot 2`] = `<MyLoader />`;

exports[`should match the snapshot 3`] = `
<div
  className="search_area"
>
  <Container
    className="primary-container"
  >
    <form
      className="search_form"
      onSubmit={[Function]}
    >
      <Icon
        as="i"
        className="search_icon"
        name="search"
        size="large"
      />
      <input
        className="search_input"
        onChange={[Function]}
        placeholder="Search for a Movie, Tv Show or Person"
        type="text"
      />
    </form>
  </Container>
</div>
`;

exports[`should match the snapshot with error 1`] = `
<div
  className="search_area"
>
  <Container
    className="primary-container"
  >
    <form
      className="search_form"
      onSubmit={[Function]}
    >
      <Icon
        as="i"
        className="search_icon"
        name="search"
        size="large"
      />
      <input
        className="search_input"
        onChange={[Function]}
        placeholder="Search for a Movie, Tv Show or Person"
        type="text"
      />
    </form>
  </Container>
</div>
`;

exports[`should match the snapshot with isSearchResultsVisible 1`] = `
<div
  className="search_area"
>
  <Container
    className="primary-container"
  >
    <form
      className="search_form"
      onSubmit={[Function]}
    >
      <Icon
        as="i"
        className="search_icon"
        name="search"
        size="large"
      />
      <input
        className="search_input"
        onChange={[Function]}
        placeholder="Search for a Movie, Tv Show or Person"
        type="text"
      />
    </form>
    <ConnectFunction
      result={Array []}
    />
  </Container>
</div>
`;

exports[`should match the snapshot with loading true 1`] = `<MyLoader />`;
